### DISCUSSION TITLE
LIDAR OPTICAL FLOW Inav7+

### DISCUSSION THREAD

**Discussion by lebolk** [iNavFlight/inav#10126] [Q&A] [answered]
Hi All !
I'm using LIDAR OPFLOW MTF-01 with Inav 7.1
I've also connected a GPS with MAG

Is there a way to use both at the same time for POSHOLD / ALTHOLD ?
Maybe to force using LIDAR from the ground to 7m... then using GPS for higher altitudes ?
Or mixing data to be even more precise ?

Thanks for your help !

**Comment by sem-fly** [comment]
Hello! Did you manage to increase the lidar height? What should I do to keep the height at 5 meters? I can't do it about 1 meter higher now.

**Reply by sensei-hacker** [reply]
> Hello! Did you manage to increase the lidar height? What should I do to keep the height at 5 meters? I can't do it about 1 meter higher now.

Set inav_max_surface_altitude and nav_max_terrain_follow_alt to the maximum distance (in centimeters) that your rangefinder is reliably accurate.  The default for nav_max_terrain_follow_alt is 100 centimeters (1 meter).

**Comment by sensei-hacker** [comment, marked as answer]
> Hi All ! I'm using LIDAR OPFLOW MTF-01 with Inav 7.1 I've also connected a GPS with MAG
> 
> Is there a way to use both at the same time for POSHOLD / ALTHOLD ? Maybe to force using LIDAR from the ground to 7m... then using GPS for higher altitudes ? Or mixing data to be even more precise ?
> 
> Thanks for your help !

This is an old question, but:

![surface_mode](https://github.com/user-attachments/assets/f0cd7ba7-c98b-45f2-bb08-4b1332b101c8)

Surface mode flies the orange line, distance to the nearest obstacle seen by the rangefinder.
Regular altitude mode flies the purple line, using the GPS and baro.

Note you can't fly the purple line and the orange line at the same time.
Therefore, mixing the purple line with the orange line "to be more precise" would be kinda like mixing English and Japanese, like this:

Emgi xo iingnlo i msheru an idi jamep na saets ehrmeisauu.

It would result in gibberish.

**Reply by MohamedNour1452** [reply]
I have another question, can I make the optical flow sensor handle the altitude hold and poshold as long as it's in it's working range and let the gps handle the rest after it goes above optical flow sensor range ?

**Reply by sensei-hacker** [reply]
You can use a switch on your radio to switch between surface mode, which flies the orange line using the rangefinder, or standard altitude mode which flies the purple line using the GPS and baro. There is no mode which will somehow fly both lines at once.


![image](https://github.com/user-attachments/assets/86e11d07-712f-4768-8bda-ca581f19f669)

You have to decide - do you want to be two meters above the hilltop, or two meters above the arming point?  You can't be in both places at once, so they are two different modes.

**Reply by MohamedNour1452** [reply]
thanks

**Reply by alex-441** [reply]
Hi! Does your answer mean that we should NOT activate ALTHOLD modifier to use SURFACE modifier with OpFlow?

**Reply by PlaidPiper** [reply]
Could some if/then logic be implemented here for a hybrid mode? Like... If baro/gps altitude is lower than surface mode altitude then trust range finder? Basically something to ensure it always fails to a safer mode?

**Reply by sensei-hacker** [reply]
> Hi! Does your answer mean that we should NOT activate ALTHOLD modifier to use SURFACE modifier with OpFlow?

It does not mean that.  But you should activate the Surface modifier FIRST. You can't switch your altitude reference while you're in an altitude-controlled mode.

**Reply by sensei-hacker** [reply]
> Could some if/then logic be implemented here for a hybrid mode? Like... If baro/gps altitude is lower than surface mode altitude then trust range finder? Basically something to ensure it always fails to a safer mode?

You can implement any logic you want, but be careful -
10 MSL may be much higher than 10 AGL, or it may be much lower - and that may change suddenly at any second.

Consider the sequence of events if you fly over your house.  You're 30 feet above the surface, then suddenly 8 feet above it. Then a few seconds later, the ground is 30 feet away again. Similarly for holes.

Be mindful that you may not want to dive down into a manhole. Also you don't want to be like the guy who set up logic that prevented him from landing. 😁

**Reply by PlaidPiper** [reply]
Apologies, my background is in infrastructure automation, and this is my literal first drone, so maybe I'm overstepping my reach.. I'd love to jump in and develop some sort of fuzzy logic modes, or custom modes that integrate decision trees of if/then statements to result in safe mode, but I most likely lack all the experience required to do so. I'll check back with you all in a few... months... years...  when I've caught up? 

Btw; Thank you guys for all you do, this has been the first exciting hobby that has energized me like this for years!

**Reply by sensei-hacker** [reply]
> Apologies, my background is in infrastructure automation

Depending what exactly "infrastructure automation" means, the programmable logic controller provided in Configurator may seem more familiar to you than it does to most people. Or not, just depending.

**Reply by sensei-hacker** [reply]
Ps George- I just noticed your name. I know someone with that name. Does "TEEX" mean anything to you?

**Comment by PlaidPiper** [comment]
Holy crap it does. Who is this?
